File: clients/ruby/lib/samsara_sdk/publisher.rb
require 'json'
require 'zlib'
require 'stringio'
require 'net/http'
require 'samsara_sdk/config'

# Samsara SDK
module SamsaraSDK
  # Physical connector that Publishes messages to Samsara Ingestion API.
  class Publisher
    # Sends message to Ingestion API.
    #
    # @param data [Array<Hash>] List of events.
    # @return [Boolean] Success or failure of HTTP POST call.
    def post(data)
      url = URI.parse(Config.get[:url].chomp('/') + Config::API_PATH)
      http = Net::HTTP.new(url.host, url.port)
      http.use_ssl = url.scheme == 'https'
      http.read_timeout = Config.get[:send_timeout_ms] / 1000
      request = Net::HTTP::Post.new(url, headers)
      request.body = prepare(data)
      http.request(request).instance_of? Net::HTTPAccepted
    rescue RuntimeError
      FALSE
    end

    private

    # Wraps data in JSON and optionally compresses it.
    #
    # @param data [Array<Hash>] Data to prepare.
    # @return [String] Prepared data.
    def prepare(data)
      data = JSON.generate data
      send(Config.get[:compression], data)
    end

    # Gzip wrapper for data.
    #
    # @param data [String] Data to wrap.
    # @return [String] Gzipped data.
    def gzip(data)
      wio = StringIO.new('w')
      w_gz = Zlib::GzipWriter.new(wio)
      w_gz.write(data)
      w_gz.close
      wio.string
    end

    # None wrapper for data.
    #
    # @param data [String] Data to wrap.
    # @return [String] Original data.
    def none(data)
      data
    end

    # Helper method to generate HTTP request headers for Ingestion API.
    #
    # @return [Hash] headers
    def headers
      {
        'Accept' => 'application/json',
        'Content-Type' => 'application/json',
        'Content-Encoding' => Config.get[:compression] == :gzip ? 'gzip' : 'identity',
        Config::PUBLISHED_TIMESTAMP_HEADER => Config.timestamp.to_s
      }
    end
  end
end

File: clients/go/publisher.go
package client

import (
	"bytes"
	"compress/gzip"
	"encoding/json"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// Samsara specific HTTP Header.
const PUBLISHED_TIMESTAMP_HEADER = "X-Samsara-publishedTimestamp"

// Samsara Ingestion API endpoint.
const API_PATH = "/v1/events"

// Interface of publishing data.
// Used mainly for test purpopses.
type IPublisher interface {
	Post(data []Event) bool
}

// Physical connector that Publishes messages to Samsara Ingestion API.
type Publisher struct {
	config Config
}

// Sends message to Ingestion API.
func (p *Publisher) Post(data []Event) bool {
	jsonData, err := json.Marshal(data)
	if err != nil {
		return false
	}

	var buffer *bytes.Buffer
	if p.config.compression == "gzip" {
		buffer = gzipWrap(jsonData)
	} else {
		buffer = noneWrap(jsonData)
	}

	req, _ := http.NewRequest("POST", strings.Trim(p.config.url, "/")+API_PATH, buffer)
	p.setHeaders(req)

	client := &http.Client{
		Timeout: time.Duration(p.config.send_timeout_ms) * time.Millisecond,
	}
	resp, err := client.Do(req)
	if err != nil {
		return false
	}

	if resp.StatusCode == 202 {
		return true
	} else {
		return false
	}
}

// Helper method to generate HTTP request headers for Ingestion API.
func (p *Publisher) setHeaders(req *http.Request) {
	var compression string
	if p.config.compression == "gzip" {
		compression = "gzip"
	} else {
		compression = "identity"
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Content-Encoding", compression)
	req.Header.Set(PUBLISHED_TIMESTAMP_HEADER, strconv.FormatInt(timestamp(), 10))
}

// Gzip wrapper for data.
func gzipWrap(data []byte) *bytes.Buffer {
	var buffer bytes.Buffer
	gz := gzip.NewWriter(&buffer)
	defer gz.Close()
	gz.Write(data)
	return &buffer
}

// None wrapper for data.
func noneWrap(data []byte) *bytes.Buffer {
	return bytes.NewBuffer(data)
}
